### DISCUSSION TITLE
How to query SEQ signals for data through sql client api?

### DISCUSSION THREAD

**Discussion by Krish75** [datalust/seq-tickets#1241] [General]
Hi,

I am using sql client api to access SEQ log data. I have also created shared signals.
I am seeing my query is too slow to get the data from SEQ server. Any help to improve query perfromance.

Can I use signals as view like "stream" to query the data?

using (var connection = new SeqConnection(_seqSettings.ServerUrl, _seqSettings.ApiKey))
            {
   var r = connection.Data.QueryAsync(query).Result
}

**Comment by nblumhardt** [comment]
Hi! Thanks for getting in touch.

Yes, the `QueryAsync` function accepts a `signal` argument that can supply the ids of signals to use:

```csharp
using (var connection = new SeqConnection(_seqSettings.ServerUrl, _seqSettings.ApiKey))
{
    var signal = SignalExpressionPart.Signal("signal-12345");
    var r = await connection.Data.QueryAsync(query, signal);
}
```

You can get the ids of signals (like `"signal-12345"` above) by retrieving signals with `connection.Signals.ListAsync(shared: true)` or similar.

To combine signals, so that only the events matching both of them are returned, use `Intersect`, e.g.:

```csharp
    var signal = SignalExpressionPart.Intersect(
        SignalExpressionPart.Signal("signal-12345"),
        SignalExpressionPart.Signal("signal-67890"));
```

Hope this helps, please let me know if you need any further information.

Best regards,
Nick

**Reply by CrossBound** [reply]
How can I build a "OR" combination of signals? For example, i want to see logs in the "warning" OR "error" signals?

**Reply by nblumhardt** [reply]
Hi @CrossBound; `SignalExpressionPart.Union()` implements the "or" operator; e.g.:

```csharp
    var signal = SignalExpressionPart.Union(
        SignalExpressionPart.Signal("signal-12345"),
        SignalExpressionPart.Signal("signal-67890"));
```
Does this do the job for you?

**Reply by CrossBound** [reply]
Yes, that helped. Thank you
